core/logger/audit/audit_logger.go

```diff
@@ -0,0 +1,231 @@
+}
+
+func NewAuditLoggerConfig(forwardToUrl string, isDev bool, jsonWrapperKey string, encodedHeaders string) (AuditLoggerConfig, error) {
+	environment := "production"
+	if isDev {
+		environment = "develop"
+	}
+
+	if forwardToUrl == "" {
+		return AuditLoggerConfig{}, errors.Errorf("No forwardToURL provided")
+	}
+
+	// Split and prepare optional service client headers from env variable
+	headers := []serviceHeader{}
+	if encodedHeaders != "" {
+		headerLines := strings.Split(encodedHeaders, "\\")
+		for _, header := range headerLines {
+			keyValue := strings.Split(header, "||")
+			if len(keyValue) != 2 {
+				return AuditLoggerConfig{}, errors.Errorf("Invalid headers provided for the audit logger. Value, single pair split on || required, got: %s", keyValue)
+			}
+			headers = append(headers, serviceHeader{
+				header: keyValue[0],
+				value:  keyValue[1],
+			})
+		}
+	}
+
+	return AuditLoggerConfig{
+		forwardToUrl:   forwardToUrl,
+		environment:    environment,
+		jsonWrapperKey: jsonWrapperKey,
+		headers:        headers,
+	}, nil
+}
+
+type AuditLoggerService struct {
+	logger          logger.Logger
+	enabled         bool
+	serviceURL      string
+	serviceHeaders  []serviceHeader
+	jsonWrapperKey  string
+	environmentName string
+	hostname        string
+	localIP         string
+	loggingChannel  chan WrappedAuditLog
+}
+
+// Configurable headers to include in POST to log service
+type serviceHeader struct {
+	header string
+	value  string
+}
+
+type WrappedAuditLog struct {
+	eventID EventID
+	data    Data
+}
+
+// NewAuditLogger returns a buffer push system that ingests audit log events and
+// asynchronously pushes them up to an HTTP log service.
+// Parses and validates the AUDIT_LOGS_* environment values and returns an enabled
+// AuditLogger instance. If the environment variables are not set, the logger
+// is disabled and short circuits execution via enabled flag.
+func NewAuditLogger(logger logger.Logger, config *AuditLoggerConfig) (AuditLogger, error) {
+	if config == nil {
+		logger.Info("Audit logger configuration is nil. Cannot start audit logger subsystem and audit events will not be captured.")
+		return &AuditLoggerService{}, nil
+	}
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		return &AuditLoggerService{}, errors.Errorf("Audit Log initialization error - unable to get hostname: %s", err)
+	}
+
+	loggingChannel := make(chan WrappedAuditLog, bufferCapacity)
+
+	// Finally, create new auditLogger with parameters
+	auditLogger := AuditLoggerService{
+		logger:          logger.Helper(1),
+		enabled:         true,
+		serviceURL:      config.forwardToUrl,
+		serviceHeaders:  config.headers,
+		jsonWrapperKey:  config.jsonWrapperKey,
+		environmentName: config.environment,
+		hostname:        hostname,
+		localIP:         getLocalIP(),
+		loggingChannel:  loggingChannel,
+	}
+
+	// Start our go routine that will receive logs and send them out to the
+	// configured service.
+	go auditLogger.auditLoggerRoutine()
```

Comment:
Generally speaking, we don't like to silently spin off goroutines in a `NewXXX` function. Please implement `service.ServiceCtx` interface instead.